Restructure Local Development Docs + Add "Customize Images" Content (#82)
astronomer/cloud-docs: docs/develop-project.md

- Use the Airflow CLI
- Add environment variables locally

## Prerequisites

To develop locally, you need:

- An existing [Astronomer project](create-project).
- [The Astronomer CLI](install-cli)
- [Docker](https://www.docker.com/products/docker-desktop)

## Build and Run a Project

To run your Astronomer project locally, run `astro dev start`. This command builds your project and spins up 3 Docker containers on your machine, each for a different Airflow component:

- **Postgres:** Airflow's metadata database
- **Webserver:** The Airflow component responsible for rendering the Airflow UI
- **Scheduler:** The Airflow component responsible for monitoring and triggering tasks

Once the project builds, you can access the Airflow UI by going to `http://localhost:8080/` and logging in with `admin` for both your username and password. You can also access your Postgres database at `localhost:5432/postgres`.

> **Note:** If you see Error: cannot start, project already running when you run this command, it means your local Airflow environment is already running your project. If there are changes you'd like to apply to your project, see Rebuild a Project.

### Restart a Local Project

> To restart a locally running Astronomer project, run the following two commands:

Author: I like "Restart Your Local Environment"! Don't think it's wordy.

```sh
$ astro dev stop
$ astro dev start
```

These commands rebuild your image and restart the Docker containers running on your local machine with that new image. Alternatively, you can run just `astro dev stop` to stop your Docker containers without restarting or rebuilding your project.

## Rebuild a Project

> All Astronomer projects require you to specify a Debian-based Astronomer Runtime image in a `Dockerfile`. When you run your project locally or on Astronomer Cloud, all of your DAG code, packages, and configurations are built into this image.

Author: "based on Astronomer Runtime?"

> All Astronomer projects require you to specify a Debian-based Astronomer Runtime image in a `Dockerfile`. When you run your project locally or on Astronomer Cloud, all of your DAG code, packages, and configurations are built into this image.

Reviewer: Ya I think you got it here

Once you've saved these packages in your project files, [rebuild your project](develop-locally#rebuild-a-project).

### Confirm your package was installed

If you added `pymongo` to your `requirements.txt` file, for example, you can confirm that it was properly installed by running a `docker exec` command into your Scheduler:

1. Run `docker ps` to identify the 3 running docker containers on your machine
2. Copy the container ID of your Scheduler container
3. Run the following:

```
docker exec -it <scheduler-container-id> pip freeze | grep pymongo

pymongo==3.7.2
```

## Configure `airflow_settings.yaml` (Local Development Only)

When you first initialize a new Astronomer project, a file called `airflow_settings.yaml` is automatically generated. With this file, you can configure and programmatically generate Airflow [Connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html), [Pools](https://airflow.apache.org/docs/apache-airflow/stable/concepts/pools.html), and [Variables](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html) so that you don't have to manually redefine these objects every time you restart your project.

> As a security measure, `airflow_settings.yaml` works only in local environments. If you deploy your project to a Deployment, the values in this file will not be pushed. To more easily manage your Airflow objects on Astronomer, we recommend [configuring a secrets backend](https://www.astronomer.io/docs/cloud/stable/customize-airflow/secrets-backend).

Author: I can switch to "values", they're pretty much the same

---
sidebar_label: 'Develop a Project'
title: 'Develop Your Astronomer Project'
id: develop-project
---

import {siteVariables} from '@site/src/versions';

## Overview

This document explains the various ways you can modify and build your Astronomer project to fit your team's use case. Specifically, this guide provides instructions on how to:

- Add DAG files to your project
- Add Python and OS-level packages to your project
- Add dependencies to your project
- Run on-build commands
- Use the Airflow CLI

> - Add environment variables locally

Reviewer: Remove "Run Airflow CLI commands" here. Also, should this bullet list actually match the H2s in this doc maybe?

## Prerequisites

> To develop locally, you need:

Reviewer: ```suggestion
To customize an Astronomer project and test it locally, you need:
```

- An existing [Astronomer project](create-project).
- [The Astronomer CLI](install-cli)
- [Docker](https://www.docker.com/products/docker-desktop)

> ## Build and Run a Project

Reviewer: ```suggestion
## Build and Run a Project Locally
```

To run your Astronomer project locally, run the following command:

```sh
astro dev start
```

This command builds your project and spins up 3 Docker containers on your machine, each for a different Airflow component:

- **Postgres:** Airflow's metadata database
- **Webserver:** The Airflow component responsible for rendering the Airflow UI
- **Scheduler:** The Airflow component responsible for monitoring and triggering tasks

Once the project builds, you can access the Airflow UI by going to `http://localhost:8080/` and logging in with `admin` for both your username and password. You can also access your Postgres database at `localhost:5432/postgres`.

::: info

> The Astronomer CLI is a wrapper around [Docker Compose](https://docs.docker.com/compose/), a tool for defining and running multi-container Docker applications. If you're familiar with Docker Compose, you'll recognize that the astro dev start command is functionally equivalent to docker compose start.

Reviewer: ```suggestion
The Astronomer CLI is a wrapper around [Docker Compose](https://docs.docker.com/compose/), a tool for defining and running multi-container Docker applications. If you're familiar with Docker Compose, you'll recognize that the `astro dev start` command, for example, is functionally equivalent to `docker compose start`.
```

:::

> **Note:** If you see `Error: cannot start, project already running` when you run this command, it means your local Airflow environment is already running your project. If there are changes you'd like to apply to your project, see Rebuild a Project.

### Restart Your Local Environment

To restart your local Airflow environment, run the following two commands:

```sh
$ astro dev stop
$ astro dev start
```

These commands rebuild your image and restart the Docker containers running on your local machine with that new image. Alternatively, you can run just `astro dev stop` to stop your Docker containers without restarting or rebuilding your project.

## Make Changes to Your Project

All Astronomer projects require you to specify a Debian-based Astronomer Runtime image in a `Dockerfile`. When you run your project locally or on Astronomer Cloud, all of your DAG code, packages, and configurations are built into a Docker image based on Astronomer Runtime.

Depending on the change you're making to your Astronomer project, you might have to rebuild your image to run your changes locally.

### DAG Code Changes

> All changes made to the following files will be live in your project as soon as you save them to your code editor:

Reviewer: ```suggestion
All changes made to the following files will be live in your local Airflow environment as soon as you save them to your code editor:
```

- `dags`
- `plugins`
- `include`

Once you save your changes, refresh the Airflow UI in your browser to see them render.

### Environment Changes

All changes made to the following files require rebuilding your image:

- `packages.txt`
- `Dockerfile`
- `requirements.txt`
- `airflow_settings.yaml`

To rebuild your project after making a change to any of these files, run the following two commands:

```sh
$ astro dev stop
$ astro dev start
```

In addition to rebuilding your image, these commands restart the Docker containers running your local Airflow environment.

> > **Note:** Note: As you develop locally, it may be necessary to reset your Docker containers and metadata DB for testing purposes. To do so, run `astro dev kill` instead of `astro dev stop` when rebuilding your image. This deletes all data associated with your local Postgres metadata database, including Airflow Connections, logs, and task history.

Reviewer: I suppose we could consolidate this info up there? The `astro dev kill` stuff and maybe just link out to section above instead of re-stating astro dev start > astro dev stop.

Don't feel strongly about this one tho - up to you

:::caution
If you are storing your project in a public directory or version control tool, we recommend adding this file to your `.gitignore` or equivalent secret management service.
:::

### Add Airflow Connections, Pools, and Variables

By default, the `airflow_settings.yaml` file includes the following template:

```yaml
airflow:
  connections:
    - conn_id: my_new_connection
      conn_type: postgres
      conn_host: 123.0.0.4
      conn_schema: airflow
      conn_login: user
      conn_password: pw
      conn_port: 5432
      conn_extra:
  pools:
    - pool_name: my_new_pool
      pool_slot: 5
      pool_description:
  variables:
    - variable_name: my_variable
      variable_value: my_value
```

This template includes all possible configuration values. If you want to add another Connection, Pool, or Variable, you can copy the existing fields for the given resource and replace the default values with your own. For instance, to create another Variable, you can add its values under the existing default Variable like so:

```yaml
variables:
  - variable_name: my_variable
    variable_value: my_value
  - variable_name: my_second_variable
    variable_value: value987
```

> Once you've saved those packages in your project files, [restart your local environment](develop-project#restart-your-local-environment). When you access the Airflow UI for your project at `localhost:8080`, you should see your new resources in the **Connections**, **Pools**, and **Variables** tabs.

Reviewer: ```suggestion
Once you've saved these values in your `airflow_settings.yaml`, [restart your local environment](develop-project#restart-your-local-environment). When you access the Airflow UI for your project at `localhost:8080`, you should see these values in the **Connections**, **Pools**, and **Variables** tabs.
```

Reviewer: i don't think we're doing anything w packages here in this airflow_settings.yaml section

## Run Commands on Build

To run extra system commands when your Airflow image builds, add them to your `Dockerfile` as a `RUN` command. These commands run as the last step in the image build process.

For example, if you want to run `ls` when your image builds, your `Dockerfile` would look like this:

> <pre><code parentName="pre">{`FROM quay.io/astronomer/astro-runtime:${siteVariables.runtimeVersion}

Reviewer: Is this missing the actual `RUN ls`? This can be a next step but I also wonder if maybe we can ask the team what a common RUN command is and add it here

`}</code></pre>

> ## Add Environment Variables (Local development only)

Reviewer: ```suggestion
## Add Environment Variables (Local Development only)
```

I don't care either way but making this consistent with the above
